### PR TITLE
Add support for source filtering in search

### PR DESCRIPTION
**Pull request by thomasvnoort**
This follows the syntax described [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-source-filtering.html). 

I'm not too sure about the constructor names, so let me know if you can think of better ones!

### CODE REVIEW

File: src/Database/Bloodhound/Types.hs [hunk 1]
```diff
@@ -604,7 +605,14 @@ data Search = Search { queryBody       :: Maybe Query
                        -- default False
                      , trackSortScores :: TrackSortScores
                      , from            :: From
-                     , size            :: Size } deriving (Eq, Show)
+                     , size            :: Size
+                     , source          :: Maybe Source } deriving (Eq, Show)
+
+data Source =
+  NoSource
+  | SourceFields FieldOrFields
```
Comment (bitemyapp) [inline src/Database/Bloodhound/Types.hs]: What's `FieldOrFields` doing for us here?

File: src/Database/Bloodhound/Types.hs [hunk 2]
```diff
@@ -604,7 +605,14 @@ data Search = Search { queryBody       :: Maybe Query
...
+  | SourceIncludeExclude [Text] [Text]
```
Comment (bitemyapp) [inline src/Database/Bloodhound/Types.hs]: Can we do better on the types than a product of two `[Text]` values?

File: src/Database/Bloodhound/Types.hs (hunk 1 above)
Comment (thomasvnoort) [inline src/Database/Bloodhound/Types.hs]: Well, according to the second example in the [specification](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-source-filtering.html), you can also pass in a single field as the source filter or a list of fields. I figured this is what `FieldOrFields` models so I might as well use it :) But if I misunderstood the purpose of that datatype we change it of course.

File: src/Database/Bloodhound/Types.hs (hunk 2 above)
Comment (thomasvnoort) [inline src/Database/Bloodhound/Types.hs]: I can change it to use some newtypes:

```
[..]
  | SourceIncludeExclude Include Exclude

data Include = Include [Text]
data Exclude = Exclude [Text]
```

Would you prefer that?

Comment (bitemyapp) [inline src/Database/Bloodhound/Types.hs]: I'd go a step further:

```haskell
newtype Pattern = Pattern Text

data Include = Include [Pattern]
data Exclude = Exclude [Pattern]
```

This leaves the door open for having a better typed grammar for patterns later (an AST'ish thing) and makes the specification more self-documenting.

**Comment by thomasvnoort** [comment]
I changed the `Source` datatype per your suggestion and took it one step further since all fields in a source specification are patterns.

**Comment by bitemyapp** [comment]
@thomasvnoort next step, could you add a test that leverages this? I understand tests in general aren't passing but I'd like to know if this functionality breaks in the future.

**Comment by thomasvnoort** [comment]
Any chance this pull request will be merged soon? I can rebase/merge my branch but only if there's a point to it.

**Comment by bitemyapp** [comment]
@thomasvnoort this one has merge conflicts.

**Comment by thomasvnoort** [comment]
I know and I can resolve them, but only if you plan on merging this pull request. Otherwise I'm doing work for nothing.

**Comment by bitemyapp** [comment]
@thomasvnoort mate, this is the 39th pull request to this repository.

I was waiting on you to fix the merge conflict. It's the responsibility of the contributor to push the ball forward, ping people, make sure the merge is still clean and the tests still pass. On the other PR I had actually already bounced the tests after fixing them (and they were broken for your changes then too) but you didn't notice and didn't fix them.

**Comment by thomasvnoort** [comment]
No need to be passive aggressive. The reason I asked is the fact that this PR was ready to be merged without conflicts a while ago. I'd be happy to resolve the conflicts that were introduced in the mean time.

**Comment by thomasvnoort** [comment]
I merged master again into this PR, everything seems to pass. Would you be so kind as to merge it? 

PR #40 is up to date too but has one conflict with this PR. I can rebase it now or fix it once this PR is merged into master, whatever you like.

**Comment by thomasvnoort** [comment]
I resolved the conflicts and the tests are passing, this seems ready to be merged.

**Comment by bitemyapp** [comment]
@thomasvnoort thank you!

**Comment by thomasvnoort** [comment]
Thank you!

**Pull request merged**